openeogeotrellis/integrations/identity.py

```diff
@@ -0,0 +1,130 @@
+    class method `instance() to get the singleton object`.
+    """
+    _TOKEN_EXPIRES_IN_SECONDS = 48 * 3600
+    _SINGLETON: Optional[IDPTokenIssuer] = None
+
+    def __init__(self):
+        if self.__class__._SINGLETON is not None:
+            raise RuntimeError("IDPTokenIssuer is meant as a singleton")
+        self._IDP_DETAILS: Optional[IDPDetails] = None
+        self.__class__._SINGLETON = self
+
+    @classmethod
+    def instance(cls) -> "IDPTokenIssuer":
+        if cls._SINGLETON is not None:
+            return cls._SINGLETON
+        else:
+            return cls()
+
+    def _reload_idp_details_if_needed(self) -> None:
+        idp_details_file = get_backend_config().openeo_idp_details_file
+        register_reload = get_file_reload_register_func_if_changed(idp_details_file)
+        if register_reload is None:
+            return
+        try:
+            self._IDP_DETAILS = IDPDetails.from_file(idp_details_file)
+            register_reload()
+        except AssertionError as ae:
+            logger.fatal(f"Invalid config staged for idp: {ae}")
+        except Exception as e:
+            logger.warning(f"Could not reload IDP details from {idp_details_file} due to {e}")
+
+    def get_job_token(self, sub_id: str, user_id: str, job_id: str) -> Optional[str]:
+        """
+        Get a JWT token that can be used by a job's execution environment to proof it is
+        really the job it claims to be and that it is issued by a specific user.
+        """
+        self._reload_idp_details_if_needed()
+        if self._IDP_DETAILS is None:
+            return None
+        else:
+            idp_details = self._IDP_DETAILS
+            now = utcnow()
+            return jwt.encode(
+                {
+                    "sub": sub_id,
+                    "exp": now + timedelta(seconds=self._TOKEN_EXPIRES_IN_SECONDS),
+                    "nbf": now,
+                    "iss": idp_details.issuer,
+                    "iat": now,
+                    "https://aws.amazon.com/tags": {
+                        "principal_tags": {
+                            "user_id": [user_id],
+                            "job_id": [job_id],
+                        },
+                        "transitive_tag_keys": [
+                            "user_id",
+                            "job_id",
+                        ]
+                    }
+                },
+                idp_details.private_key,
+                algorithm="RS256"
+            )
+
+
+IDP_TOKEN_ISSUER = IDPTokenIssuer.instance()
+
+
+if __name__ == '__main__':
+    aws_config_dir = os.environ.get("AWS_CONFIG_DIR", "/opt/spark/work-dir")
+    aws_config_path = Path(aws_config_dir)
+    aws_token_filename = os.environ.get("AWS_TOKEN_FILENAME", "token")
+    aws_config_path.mkdir(parents=True, exist_ok=True)
+    with open(aws_config_path.joinpath(aws_token_filename), 'w') as token_file:
+        token_file.write(IDP_TOKEN_ISSUER.get_job_token("0", "none"))
```

Comment:
If OK I want to leave it in there for allowing troubleshooting of drivers when needed. Basically running this python file on the driver should store a token in a local file which would not influence operations but would allow to do some basic tests with the generated token. (e.g. making sure infra dependencies are correct).

Reply:
I wasn't aware of this behavior but it does make sense.